### ISSUE TITLE
Crash in MVKDescriptorSetLayoutBinding::bind()

### ISSUE THREAD

**Issue by alexey-lysiuk** [KhronosGroup/MoltenVK#651]
I encountered a pretty stable way to crash MoltenVK which I cannot reproduce with other Vulkan implementations (Windows and Linux, NVIDIA/AMD/Intel hardware).

Here is a callstack.
```
MVKDescriptorSetLayoutBinding::bind(MVKCommandEncoder*, MVKDescriptorBinding&, MVKShaderResourceBinding&, MVKVector<unsigned int>&, unsigned int*) + 1122
MVKDescriptorSetLayout::bindDescriptorSet(MVKCommandEncoder*, MVKDescriptorSet*, MVKShaderResourceBinding&, MVKVector<unsigned int>&, unsigned int*) + 121
MVKPipelineLayout::bindDescriptorSets(MVKCommandEncoder*, MVKVector<MVKDescriptorSet*>&, unsigned int, MVKVector<unsigned int>&) + 125
MVKCommandBuffer::submit(MVKQueueCommandBufferSubmission*) + 434
MVKQueueCommandBufferSubmission::execute() + 67
invocation function for block in MVKQueue::submit(MVKQueueSubmission*)
```

The problem is the steps to reproduce are rather complex.
Unfortunately, I unable to create a simple test case for it.

In short, you need to run [GZDoom](https://zdoom.org/downloads) with Vulkan renderer turned on and with a particular mod.
To simplify the process, I created the following shell script.
```bash
#!/bin/sh

# Abort on error
set -o errexit

# Setup work directory
mkdir moltenvk_crash
cd moltenvk_crash

# Download the latest release of GZDoom
curl -LO https://github.com/coelckers/gzdoom/releases/download/g4.1.3/gzdoom-bin-4-1-3.dmg

# Download and extract game data
curl -LO https://github.com/freedoom/freedoom/releases/download/v0.11.3/freedoom-0.11.3.zip
unzip freedoom-0.11.3.zip freedoom-0.11.3/freedoom1.wad

# Download Vanilla Essence mod
curl -o "Vanilla Essence 4.3.pk3" "https://forum.zdoom.org/download/file.php?id=35328"

# Mount GZDoom disk image and verify its digital signature
open -W gzdoom-bin-4-1-3.dmg
pkgutil --check-signature /Volumes/GZDoom-g4.1.3/GZDoom.app

# Run GZDoom
/Volumes/GZDoom-g4.1.3/GZDoom.app/Contents/MacOS/gzdoom -iwad freedoom-0.11.3/freedoom1.wad -file "Vanilla Essence 4.3.pk3" +vid_enablevulkan 1 +map e1m1 +bind q ve_toggle
```

When game is loaded, press `Q` button several times. Usually, it crashes after a few presses.

The crash is caused by mismatch in bindings of descriptor set layout and bindings of descriptor set itself. So far, I managed to track it to [this line](https://github.com/KhronosGroup/MoltenVK/blob/17496c30dfcefc5c48c5d6eae10b5721af833541/MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm#L981).
From what I figure out during debugging of this crash, it's possible that pointer to layout remains unchanged while layout itself is different.
Removal of layout pointer caching "fixes" the crash. Disable pooling of MVKDescriptorSet objects "fixes" it as well.
Synchronous queue submit was disabled, although the crash just moves to the main thread otherwise.

If you would like to check the crash in details, it's possible to build [GZDoom](https://github.com/coelckers/gzdoom) from sources and to run it from Xcode, using CMake generated project or even MoltenVK project.

Please let me know if you any additional information regarding this issue.

**Comment by alexey-lysiuk** [comment]
To catch the problem earlier, you can add the following condition.
```diff
diff --git a/MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm b/MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm
index 1396bac..0f30afa 100644
--- a/MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm
+++ b/MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm
@@ -988,6 +988,10 @@
 			_bindings.emplace_back(this, &layout->_bindings[i]);
 		}
 	}
+	else if (layout->_bindings.size() != _bindings.size())
+	{
+		__builtin_trap();
+	}
 }
 
 
```
This helps to stop in the debugger when layout mismatch happens at the place where it's set.

Also, steps to reproduce can be simplified a bit by running GZDoom without any mods, enabling internal bloom postprocess, and switching between windowed and fullscreen by pressing `Command + F`.
In this case, command line will be the following
```bash
/Volumes/GZDoom-g4.1.3/GZDoom.app/Contents/MacOS/gzdoom -iwad freedoom-0.11.3/freedoom1.wad +vid_enablevulkan 1 +gl_bloom 1 +map e1m1
```

Setting `MVK_CONFIG_SYNCHRONOUS_QUEUE_SUBMITS` environment variable to `1` before running GZDoom will disable asynchronous queue submission without recompilation.

**Comment by alexey-lysiuk** [comment]
The problem is boils down to a layout handle created by `vkCreateDescriptorSetLayout()` function can have the same value (address or pointer) again.
In other words, create -> destroy -> create sequence may produce the same handle while `VkDescriptorSetLayoutCreateInfo` contents are different between two calls.

At the same time, this layout handle is used as unique identifier in [MVKDescriptorSet::setLayout()](https://github.com/KhronosGroup/MoltenVK/blob/251e5200c4b87d06c05e5c3397f5a3ca68424a50/MoltenVK/MoltenVK/GPUObjects/MVKDescriptorSet.mm#L980-L991) function.
Because of descriptor set pooling, an existing object with non-null layout pointer can acquired from the pool.
However, there is no guaranty that the stored pointer points to the same layout object.
Like with GZDoom, it may crash or may use wrong descriptor set layout.

If aforementioned is correct, a raw layout pointer cannot be not used as unique key for layout caching reliably.
I guess such key should include additional information like bindings.

**Comment by billhollings** [comment]
@alexey-lysiuk 

This has hopefully been fixed in PR #553 

This solution destroys the descriptor set pool associated with a descriptor set layout when that layout is destroyed. In addition to fixing the specific reuse-mismatch issue described here, that should also help with the memory build up that would occur in an app that keeps destroying and recreating layout.

Can you test it out in your environment, and if it fixes the problem, close this issue?

**Comment by alexey-lysiuk** [comment]
Verified with 4e4bf617c6819de1f3b05fa80f20fdd520b09594 that the issue was resolved. Thank you.
